sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+
+    private final UnboundedMqttSource source;
+
+    private transient MqttClient client;
+    private transient byte[] current;
+    private transient Instant currentTimestamp;
+    private transient Instant watermark;
+    private transient BlockingQueue<KV<MqttMessage, Instant>> queue;
+
+    private UnboundedMqttReader(UnboundedMqttSource source) {
+      this.source = source;
+      this.current = null;
+      this.queue = new LinkedBlockingQueue<>();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOGGER.info("Starting MQTT reader");
+      Read spec = source.spec;
+      try {
+        client = spec.mqttConnectionConfiguration().getClient();
+        client.setCallback(new MqttCallback() {
+          @Override
+          public void connectionLost(Throwable cause) {
+            LOGGER.warn("MQTT connection lost", cause);
+          }
+
+          @Override
+          public void messageArrived(String topic, MqttMessage message) throws Exception {
+            LOGGER.info("Message arrived");
+            currentTimestamp = Instant.now();
+            queue.put(KV.of(message, currentTimestamp));
+          }
+
+          @Override
+          public void deliveryComplete(IMqttDeliveryToken token) {
+            // nothing to do
+          }
+        });
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      LOGGER.info("Taking from the pending queue ({})", queue.size());
+      try {
+        KV<MqttMessage, Instant> message = queue.take();
+        current = message.getKey().getPayload();
+        watermark = message.getValue();
+        if (current == null) {
+          return false;
+        } else {
+          return true;
+        }
+      } catch (InterruptedException e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public void close() throws IOException {
+      LOGGER.info("Closing MQTT reader");
+      try {
+        if (client != null) {
+          try {
+            client.disconnect();
+          } finally {
+            client.close();
+          }
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return watermark;
+    }
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return new UnboundedSource.CheckpointMark() {
+        @Override
+        public void finalizeCheckpoint() throws IOException {
+          // nothing to do
+        }
+      };
+    }
+
+    @Override
+    public byte[] getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
```

Comment:
getCurrent() and getCurrentTimestamp() should be placed next to each other for better readability. Currently they are separated by getCurrentSource().